### PR TITLE
ARROW-16064: [Java] [C++] [FlightRPC] Add missing column metadata for type name on FlightSQL

### PR DESCRIPTION
**Pull request by iurysalino**
This adds a new column metadata field ARROW:FLIGHT:SQL:TYPE_NAME to FlightSQL resultset schemas.

This improvement allows datasources to specify custom data type names on JDBC driver.

### CODE REVIEW

File: format/FlightSql.proto [hunk 1]
```diff
@@ -1118,6 +1118,7 @@ message CommandGetDbSchemas {
  *  - ARROW:FLIGHT:SQL:CATALOG_NAME      - Table's catalog name
  *  - ARROW:FLIGHT:SQL:DB_SCHEMA_NAME    - Database schema name
  *  - ARROW:FLIGHT:SQL:TABLE_NAME        - Table name
+ *  - ARROW:FLIGHT:SQL:TYPE_NAME         - Type name
```
[the remark this answers is not in the document]
Comment (lidavidm) [reply format/FlightSql.proto:1121]: (Does your team want to start the discussion on the mailing list?)

[the remark this answers is not in the document]
Comment (iurysalino) [reply format/FlightSql.proto:1121]: We are about to start it.

File: format/FlightSql.proto [hunk 2]
```diff
@@ -1491,6 +1493,7 @@ message TicketStatementQuery {
  *    - ARROW:FLIGHT:SQL:CATALOG_NAME      - Table's catalog name
  *    - ARROW:FLIGHT:SQL:DB_SCHEMA_NAME    - Database schema name
  *    - ARROW:FLIGHT:SQL:TABLE_NAME        - Table name
+ *    - ARROW:FLIGHT:SQL:TYPE_NAME         - Type name
```
[the remark this answers is not in the document]
Comment (lidavidm) [reply format/FlightSql.proto:1496]: Right, are these values fixed in the standard, or are they meant to reflect implementation-specific values?

**Review by lidavidm — COMMENTED** [review]
This looks fairly straightforward, thank you.

Can this be added to the SQLite/Derby examples?

**Comment by jcralmeida** [comment]
Hey @lidavidm . I've already updated the PR with integration tests to use the new column metadata. However, some of the tests in the CI seems to be failing. I've looked through the errors and they don't seem to be related to any of our changes.

**Comment by lidavidm** [comment]
Can you please rebase? That should fix most of the gcsfs related issues.

**Comment by jcralmeida** [comment]
Thanks David, I just did it. Waiting to see the results on the tests

**Review by lidavidm — APPROVED** [review]

**Comment by jvictorhuguenin** [comment]
[b-actions](https://github.com/github-actions) crossbow submit java-jars

**Comment by jvictorhuguenin** [comment]
@github-actions crossbow submit java-jars

**Comment by jcralmeida** [comment]
Hi @lidavidm, thanks for you review and approval. What are our next steps? Will you initiate the voting?

**Comment by lidavidm** [comment]
@emkornfield or @pitrou any comments here?

File: format/FlightSql.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply format/FlightSql.proto:1121]: I see the discussion started but the vote wasn't formally initiated yet?

**Review by pitrou — APPROVED** [review]
I have no problems on the principle but the vote hasn't been made yet.

File: format/FlightSql.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (lidavidm) [reply format/FlightSql.proto:1121]: I'll start the vote now.

File: cpp/src/arrow/flight/sql/column_metadata.h [hunk 3]
```diff
@@ -78,6 +81,10 @@ class ColumnMetadata {
   /// \return The table name.
   arrow::Result<std::string> GetTableName() const;
 
+  /// \brief  Return the type name set in the KeyValueMetadata.
```
Comment (emkornfield) [inline cpp/src/arrow/flight/sql/column_metadata.h:84]: can you expand the comments here what is type name?

Comment (lidavidm) [reply cpp/src/arrow/flight/sql/column_metadata.h:84]: Ping @iurysalino @jduo would it be possible to clarify things here before merging?

Comment (pbquick) [reply cpp/src/arrow/flight/sql/column_metadata.h:84]:
```suggestion
  /// \brief  Return the data-source specific name for the data type of the column.
```

Comment (jduo) [reply cpp/src/arrow/flight/sql/column_metadata.h:84]:
```suggestion
  /// \brief  Return the data source-specific name for the data type of the column.
```

Comment (jcralmeida) [reply cpp/src/arrow/flight/sql/column_metadata.h]: I improved the doc here using James suggestion.

Comment (lidavidm) [reply cpp/src/arrow/flight/sql/column_metadata.h]: Can the clarification be done in the format itself as well?

Comment (jcralmeida) [reply cpp/src/arrow/flight/sql/column_metadata.h]: Sure, I already did it

**Review by lidavidm — APPROVED** [review]

**Pull request closed without merging**